File: sast-java/src/main/java/com/sast/astbenchmark/case_language_maturity/completeness/single_app_tracing/function_call/higher_order_function/higher_order_function_002_F.java
package com.sast.astbenchmark.case_language_maturity.completeness.single_app_tracing.function_call.higher_order_function;

import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.io.IOException;
import java.util.function.Consumer;

/**
 * Introduction 完整度->单应用跟踪完整度->函数和方法调用->高阶函数
 * Level 2
 * Date 2025-02-13
 */

// evaluation information start
// real case = false
// evaluation item = 完整度->单应用跟踪完整度->函数和方法调用->高阶函数
// bind_url = completeness/single_app_tracing/function_call/higher_order_function/higher_order_function_002_F
// evaluation information end
@RestController()
@RequestMapping("completeness/single_app_tracing/function_call/higher_order_function")
public class higher_order_function_002_F {
    public static void applyConsumer(Consumer<String> consumer, String message) {
        consumer.accept(message);
    }

    @PostMapping(value = "higher_order_function_002_F")
    public void higher_order_function_002_F(@RequestParam String __taint_src) {
        applyConsumer(higher_order_function_002_F::__safe, __taint_src);
    }

    private static void __taint_sink(Object o) {
        try {
            Runtime.getRuntime().exec(String.valueOf(o));
        } catch (IOException e) {
        }
    }

    private static void __safe(Object o) {
    }

}

File: sast-php/case/completeness/single_app_tracing/function_call/higher_order_function_002_F.php
<?php
// evaluation information start
// real case = false
// evaluation item = 完整度->单应用追踪->函数调用
// scene introduction = usort 回调仅用于比较，sink 接收安全值
// level = 2
// bind_url = completeness/single_app_tracing/function_call/higher_order_function_002_F
// evaluation information end

function higher_order_function_002_F($__taint_src) {
    $arr = ["b", "a", "c"];
    usort($arr, function($a, $b) {
        return strcmp($a, $b);
    });
    __taint_sink($arr[0]);
}

function __taint_sink($o) {
    shell_exec($o);
}

$taint_src = "taint_src_value";
higher_order_function_002_F($taint_src);
